File: src/javascript/algorithms/linkedlist/findMiddleNode.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    constructor(value) {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(value) {
        const newNode = new Node(value);
        if(!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
        return this;
    }

    findMiddleNodeWithLength() {
        let temp = this.head;
        for(let i = 0; i < Math.floor(this.length/2); i++) {
            temp = temp.next;
        }
        return temp;
    }

    findMiddleNode() {
        let slow = this.head;
        let fast = this.head;
        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }
}

const myLinkedList = new LinkedList();
myLinkedList.push(1);
myLinkedList.push(2);
myLinkedList.push(3);
myLinkedList.push(4);
myLinkedList.push(5);

myLinkedList.findMiddleNode();
myLinkedList.findMiddleNodeWithLength();

File: src/java1/algorithms/linkedlist/findMiddleNode/FindMiddleNode.java
package java1.algorithms.linkedlist.findMiddleNode;

import java1.algorithms.linkedlist.Node;

public class FindMiddleNode {

    //Floyd's Tortoise and Hare algorithm
    private static Node findMiddleNode(Node head) {
        Node slow = head;
        Node fast = head;

        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }
    public static void main(String[] args) {
          Node headNode1 = new Node(1);
          Node node2 = new Node(2);
          Node node3 = new Node(3);
          Node node4 = new Node(4);
          Node node5 = new Node(5);

          Node headNode2 = new Node(1);
          Node node6 = new Node(2);
          Node node7 = new Node(3);
          Node node8 = new Node(4);
          Node node9 = new Node(5);
          Node node10 = new Node(6);


          headNode1.setNext(node2);
          node2.setNext(node3);
          node3.setNext(node4);
          node4.setNext(node5);

          headNode2.setNext(node6);
          node6.setNext(node7);
          node7.setNext(node8);
          node8.setNext(node9);
          node9.setNext(node10);

          System.out.println(findMiddleNode(headNode1).value);
          System.out.println(findMiddleNode(headNode2).value);
    }
}

